bigframes/session/_io/bigquery/read_gbq_table.py

```diff
@@ -0,0 +1,386 @@
+        raise
+
+
+def _check_index_uniqueness(
+    bqclient: bigquery.Client,
+    ibis_client: ibis.BaseBackend,
+    table: ibis_types.Table,
+    index_cols: List[str],
+    api_name: str,
+) -> bool:
+    distinct_table = table.select(*index_cols).distinct()
+    is_unique_sql = f"""WITH full_table AS (
+        {ibis_client.compile(table)}
+    ),
+    distinct_table AS (
+        {ibis_client.compile(distinct_table)}
+    )
+
+    SELECT (SELECT COUNT(*) FROM full_table) AS `total_count`,
+    (SELECT COUNT(*) FROM distinct_table) AS `distinct_count`
+    """
+    job_config = bigquery.QueryJobConfig()
+    job_config.labels["bigframes-api"] = api_name
+    results = bqclient.query_and_wait(is_unique_sql, job_config=job_config)
+    row = next(iter(results))
+
+    total_count = row["total_count"]
+    distinct_count = row["distinct_count"]
+    return total_count == distinct_count
+
+
+def _get_primary_keys(
+    table: bigquery.table.Table,
+) -> List[str]:
+    """Get primary keys from table if they are set."""
+
+    primary_keys: List[str] = []
+    if (
+        (table_constraints := getattr(table, "table_constraints", None)) is not None
+        and (primary_key := table_constraints.primary_key) is not None
+        # This will be False for either None or empty list.
+        # We want primary_keys = None if no primary keys are set.
+        and (columns := primary_key.columns)
+    ):
+        primary_keys = columns if columns is not None else []
+
+    return primary_keys
+
+
+def get_index_cols_and_uniqueness(
+    bqclient: bigquery.Client,
+    ibis_client: ibis.BaseBackend,
+    table: bigquery.table.Table,
+    table_expression: ibis_types.Table,
+    index_col: Iterable[str] | str,
+    api_name: str,
+) -> Tuple[List[str], bool]:
+    """
+    If we can get a total ordering from the table, such as via primary key
+    column(s), then return those too so that ordering generation can be
+    avoided.
+    """
+
+    # Transform index_col -> index_cols so we have a variable that is
+    # always a list of column names (possibly empty).
+    if isinstance(index_col, str):
+        index_cols: List[str] = [index_col]
+    else:
+        index_cols = list(index_col)
+
+    # If the isn't an index selected, use the primary keys of the table as the
+    # index. If there are no primary keys, we'll return an empty list.
+    if len(index_cols) == 0:
+        index_cols = _get_primary_keys(table)
+
+        # TODO(b/335727141): If table has clustering/partitioning, fail if
+        # index_cols is empty.
+
+        # If there are primary keys defined, the query engine assumes these
+        # columns are unique, even if the constraint is not enforced. We make
+        # the same assumption and use these columns as the total ordering keys.
+        is_index_unique = len(index_cols) != 0
+    else:
+        is_index_unique = _check_index_uniqueness(
```

Comment:
This operation is also quiet expensive though.

Reply:
Good point. I want to minimize any behavior changes in this PR. Filed b/338065601 to follow-up.